src/SAML2/SignedElementTrait.php

```diff
@@ -93,28 +117,23 @@ public function addValidator(callable $function, $data): void
      */
     public function validate(XMLSecurityKey $key): bool
     {
-        if (count($this->validators) === 0) {
+        if ($this->signature === null) {
             return false;
         }
 
-        $exceptions = [];
-
-        foreach ($this->validators as $validator) {
-            $function = $validator['Function'];
-            $data = $validator['Data'];
+        $signer = $this->signature->getSigner();
+        Assert::eq(
+            $key->getAlgorithm(),
+            $this->signature->getAlgorithm(),
+            'Algorithm provided in key does not match algorithm used in signature.'
+        );
 
-            try {
-                call_user_func($function, $data, $key);
-                // we were able to validate the message with this validator.
+        // check the signature
+        if ($signer->verify($key) === 1) {
+            return true;
 
-                return true;
-            } catch (Exception $e) {
-                $exceptions[] = $e;
-            }
         }
-
-        // no validators were able to validate the message.
-        throw $exceptions[0];
+        throw new \Exception("Unable to validate Signature");
```

Comment:
I was always taught to avoid the built-in \Exception ...
Should we use one of our own exceptions here, like RuntimeException?

Reply:
Maybe even a specific exception for signature validation errors...

Reply:
I'm glad we all agree :+1: